README.md
# dinit

dinit is a simple process supervisor and init system designed to run as PID 1
inside container environments.

All dinit does is manage the startup sequence of the processes in the container, and
wait for it to exit and signal forwarding.


## Features

- startup sequence of the processes
- restart in-place without exiting the container

## Environments

- `DINIT_PORT` - Port used to receive HTTP calls from the remote, default: `8080`
- `DINIT_CONFIG ` - The Configuration of the dinit
- `DINIT_GRACEFUL_TIMEOUT` - Timeout for graceful exit, default: `30`
- `DINIT_EXIT` - Whether the dinit exit when receive signal, default: `true`
- `DINIT_LOG_FILE` - The log file to write to, default: `/data/logs/dinit/dinit.log`
- `DINIT_LOG_FIELDS` - The fileds in log, default: `APP_NAME,POD_IP,HOST_IP,NODE_NAME`

## Configuration file

Example:

```yaml
version: 1.0.0

log_level: debug

services:
  main:
    exec_start: java -server -Xms$(JVM_XMS) -Xmx$(JVM_XMX)...
    lifecycle:
      pre_start:
        exec:
          command: echo "pre_start" && sleep 3
      post_start:
        exec:
          command: echo "post_start" && sleep 3
      pre_stop:
        exec:
          command: echo "pre_stop" && sleep 3
      post_stop:
        exec:
          command: echo "post_stop" && sleep 3
    environment:
      ROUTER: benchmark
    depends_on:
      - consul
  consul:
    exec_start: consul agent -retry-join=consul -config-file /data/consul/config.json
    lifecycle:
      post_start:
        exec:
          command: while true; do sleep 1 && consul catalog services > /dev/null && exit 0 ;done
```

## How to Use

### Docker

- application dockerfile

```
FROM harbor.example.com/public/runtime:slim

WORKDIR /data/apps

ENTRYPOINT ["/urs/local/bin/dinit"]

# java application
CDM ["java", "-server", "-Xms$(JVM_XMS)", "-Xmx$(JVM_XMX)"...]

# python application
CMD ["uwsgi", "--module=runserver", "--callable=app", "--master"...]
```

### Kubernetes

- k8s deployment

```
...
Name:                   xxx-service
Namespace:              default
Labels:                 app=xxx-service
Pod Template:
   dinit:
    Image:      harbor.example.com/public/dinit:1.0.0
    Port:       <none>
    Host Port:  <none>
    Command:
      sh
      -c
    Args:
      echo -en '\n'; cp /usr/local/bin/dinit /data; ls -lh /data/dinit;
    Environment:  <none>
    Mounts:
      /data from dinit (rw)
   xxx-service:
    Image:      harbor.example.com/app/xxx-service:xxxx
    Port:       <none>
    Host Port:  <none>
    Command:
      sh
      -c
    Args:
      echo -en '\n'; cp /data/apps/springboot.jar /data; ls -lh /data/springboot.jar;
    Environment:  <none>
    Mounts:
      /data from dinit (rw)
  Containers:
   app:
    Image:      harbor.example.cn/env/jdk:1.8.0_251
    Port:       8080/TCP
    Host Port:  0/TCP
    Command:
      /data/dinit
    Args:
      java
      -server
      -Xms$(JVM_XMS)
      -Xmx$(JVM_XMX)
      -javaagent:/data/pinpoint-agent/pinpoint-bootstrap.jar
      -Dpinpoint.agentId=$(POD_IP)
      -Dpinpoint.applicationName=$(APP_NAME)
      -Dfile.encoding=UTF-8
      -Djava.awt.headless=true
      -Djava.security.egd=file:/dev/./urandom
      -Xss256k -XX:+AggressiveOpts -XX:-UseBiasedLocking -XX:MaxTenuringThreshold=15 -XX:LargePageSizeInBytes=128m -XX:+UseFastAccessorMethods -XX:MaxGCPauseMillis=20 -XX:InitiatingHeapOccupancyPercent=35 -XX:MetaspaceSize=128M
      -XX:+DisableExplicitGC -XX:+UseG1GC -XX:+PrintGCDateStamps -XX:+PrintGCDetails -Xloggc:/data/logs/gc/gc.log -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/data/logs/gc/heapdump.bin
      -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap -XX:+UseContainerSupport
      -jar
      /data/springboot.jar
      --spring.profiles.active=$(SPRING_PROFILES_ACTIVE)
      --server.tomcat.accesslog.rotate=false
      --server.tomcat.accesslog.pattern="%t  %a  %A  %r  %s  %D  %b  %I" >> /data/logs/java/console.log 2>&1
    Limits:
      cpu:     4
      memory:  8Gi
    Requests:
      cpu:      1
      memory:   8Gi
    Readiness:  http-get http://:8080/health delay=80s timeout=1s period=5s #success=1 #failure=3
    Environment:
      TZ:                           Asia/Shanghai
      APP_NAME:                     xxx-service
      APP_REGISTER_NAME:            xxx-service
      JVM_XMS:                      4096m
      JVM_XMX:                      4096m
      APP_ENV:                      <set to the key 'APP_ENV' of config map 'application-environment'>                      Optional: false
      CLOUD:                        <set to the key 'CLOUD' of config map 'application-environment'>                        Optional: false
      ROUTE:                        <set to the key 'ROUTE' of config map 'application-environment'>                        Optional: false
      DINIT_GRACEFUL_TIMEOUT:       <set to the key 'DINIT_GRACEFUL_TIMEOUT' of config map 'application-environment'>       Optional: false
      SPRING_PROFILES_ACTIVE:       <set to the key 'SPRING_PROFILES_ACTIVE' of config map 'application-environment'>       Optional: false
      ZK_HOSTS:                     <set to the key 'ZK_HOSTS' of config map 'application-environment'>                     Optional: false
      CFG_REDIS:                    <set to the key 'CFG_REDIS' of config map 'application-environment'>                    Optional: false
      POD_NAME:                      (v1:metadata.name)
      POD_NAMESPACE:                 (v1:metadata.namespace)
      POD_IP:                        (v1:status.podIP)
      NODE_NAME:                     (v1:spec.nodeName)
      HOST_IP:                       (v1:status.hostIP)
    Mounts:
      ...
   filebeat:
    Image:      harbor.example.cn/public/filebeat:latest
    Port:       <none>
    Host Port:  <none>
    Limits:
      cpu:     1
      memory:  512Mi
    Requests:
      cpu:     100m
      memory:  128Mi
    Environment:
      TZ:  Asia/Shanghai
    Mounts:
      /data/logs from xxx-service (rw)
      /etc/localtime from time (rw)
  Volumes:
    ...
```


### Service Restart In-place

#### Descriptions

The main process restarts these child processes after receiving the `restart In-place` signal


Provide two methods

one example:
```bash
kill -s SIGHUP $dinit_pid
```

another example:

```bash
curl http://localhost:$dinit_port/init/reload
```


ChangeLog.md
# Changes
## 1.3.0 (2020-12-22)
md5sum: 648624b3a21a6b0a4f12017f6ca1efa4

### Added
- [dinit] Complete lifecycle policy for services

### Changed
- [dinit] Change the log formatter to JSONFormatter from TextFormatter

### Removed

- [dinit] Synchronization time for `consul` ( using `post_stop` phase instead of `DINIT_CONSUL_LEAVE_INTERVAL` )

## 1.2.1 (2020-12-18)
md5sum: c726d41cf001f1b7b42717cbc5ccd561

### Changed
- [dinit] reduce encumbrance logs

### Fixed
- [dinit] avoid repeated shutdown procedures
- [dinit] avold using invalid code to exit when the child process receivers a siganl

## 1.2.0 (2020-12-16)
md5sum: 853dfb29ad42e1bbcf4cdd348e3eaa95

### Added
- [dinit] report information to CMDB
- [dinit] offline before consul leave

## 1.1.3 (2020-12-14)
md5sum: 7e71537a23338ee090792fa7801beefe

### Changed
- [dinit] add retry for flowController
- [dinit] `DINIT_DIRECTOR_ADDR` is requirement

## 1.1.2 (2020-12-08)
md5sum: afa78edac263317521794f4ec5f13828

### Added
- [dinit] support for extra environment variable into log fields

## 1.1.1 (2020-11-19)
md5sum: caac97cb053312177df324458b3473a4

### Fixed
- [dinit] fixed maybe cause panic with nil pointer

## 1.1.0 (2020-11-12)

md5sum: 2a99ef48d6436e2a400e49ee83703f42

## 1.1.0-alpha (2020-10-27)

md5sum: d815ac979a3e2a88b443dc4a981e93ce

### Added
- [dinit] Add version information to log
- [dinit] Support service restart in place

### Changed
- [dinit] Collect logs of child process stderr
- [dinit] Add synchronization time to `consul`

## 1.0.8 (2020-09-17)

### Fixed
- [dinit] fix the missing system environment

## 1.0.7 (2020-09-09)

### Changed
- [dinit] Support for environment

## 1.0.6 (2020-08-26)

### Changed
- [dinit] Change the log output format to json format

### Fixed
- [dinit] fix the `pre_stop` pharse did not complete exit


## 1.0.5 (2020-08-14)

### Changed
- [dinit] add log output file
- [dinit] the main process hanging when the child process return non-zero exit code

## 1.0.4 (2020-07-02)

### Changed
- [dinit] - main process exit code use sub-process exit code
- [dinit] - support main process hang through environment variable of `DINIT_EXIT`
- [dinit] - flow control, stop traffic before stopping services

## 1.0.3 (2020-06-28)

### Fixed
- [watchman] - fix synchronization of new apps


## 1.0.2 (2020-06-19)

### Added
- [watchman] - support running outside of k8s

### Changed
- [dinit] - shutdown the ingress before exiting
- [dinit] - ignore the consul check for JS applications


## 1.0.1 (2020-06-18)

### Added
- [dinit] - flow control
- [watchman] - New component, a service for synchronizing the health of the
  applications between internal and external

## 1.0.0-beta (2020-04-10)

brand new version

### Feature

- Dinit as the main process in the container
- Processes lifecycle
- Service startup dependencies


pkg/dinit/app/root.go
package app

import (
	"fmt"
	"os"

	homedir "github.com/mitchellh/go-homedir"
	"github.com/rosenlo/toolkits/log"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

var cfgFile string

const (
	TimeFormatFormat string = "2006-01-02 15:04:05.000 Z0700"
)

// dinit represents the base command when called without any subcommands
var dinit = &cobra.Command{
	Use:                "dinit",
	Short:              "Container daemon process",
	Long:               `Container daemon process`,
	DisableFlagParsing: true,
	Run:                Run,
}

// Execute adds all child commands to the root command and sets flags appropriately.
// This is called by main(). It only needs to happen once to the rootCmd.
func Execute() error {
	return dinit.Execute()
}

func init() {
	cobra.OnInitialize(initConfig)
}

// initConfig reads in config file and ENV variables if set.
func initConfig() {
	if cfgFile != "" {
		// Use config file from the flag.
		viper.SetConfigFile(cfgFile)
	} else {
		// Find home directory.
		home, err := homedir.Dir()
		if err != nil {
			log.Println(err)
			os.Exit(1)
		}

		viper.SetConfigName("dinit")
		viper.AddConfigPath(".")
		viper.AddConfigPath(home)
		viper.AddConfigPath("/etc/dinit")
	}
	viper.SetEnvPrefix("DINIT")
	// basic
	viper.SetDefault("address", ":8888")
	viper.SetDefault("command", "")
	viper.SetDefault("log_level", "debug")
	viper.SetDefault("log_file", "/data/logs/dinit/dinit.log")
	viper.SetDefault("log_fields", "APP_NAME,POD_IP,NODE_NAME")
	viper.SetDefault("services_config", "/etc/dinit/dinit.yaml")
	viper.SetDefault("http_retry", 3)
	viper.SetDefault("http_retry_time", 3)

	// process
	viper.SetDefault("graceful_timeout", 30)
	viper.SetDefault("exit", true)
	viper.SetDefault("executor", "/bin/sh")
	viper.SetDefault("executor_arg", "-ec")

	// flowcontrol
	viper.SetDefault("openresty_addr", "")
	viper.SetDefault("service_prefix_key", "nginx/upstreams/")
	viper.SetDefault("health_check_interval", 10)
	viper.SetDefault("consul_username", "")
	viper.SetDefault("consul_password", "")

	viper.AutomaticEnv() // read in environment variables that match

	// If a config file is found, read it in.
	if err := viper.ReadInConfig(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	// test config
	if viper.GetString("env") == "test" {
	}
}
